lib/remap-lang/src/expression/arithmetic.rs

```diff
@@ -45,4 +52,223 @@ impl Expression for Arithmetic {
 
         result.map(Some).map_err(Into::into)
     }
+
+    fn type_def(&self, state: &state::Compiler) -> TypeDef {
+        use value::{Constraint::*, Kind::*};
+        use Operator::*;
+
+        let constraint = match self.op {
+            Or => self
+                .lhs
+                .type_def(state)
+                .constraint
+                .merge(&self.rhs.type_def(state).constraint),
+            Multiply | Add => OneOf(vec![String, Integer, Float]),
```

Comment:
`Multiply` should check the typedef of `self.lhs` to see if we are working with `String` or `Integer` or `Float`. 

`Add` should probably check both sides, but I'm not exactly sure what should happen if both are different?

Reply:
Ah yes, I forgot to finish this impl. I'll do so in a follow-up PR 👍